File: swift/Sources/Schemas/AttachmentView.swift
import Foundation

public struct AttachmentView: Codable, Hashable, Sendable {
    public let capabilities: [String]
    public let cleanupAt: Nullable<Double>
    public let cleanupStatus: AttachmentViewCleanupStatus
    public let environmentId: String
    public let id: String
    public let journalId: String
    /// Renewal TTL; anchor only ready/heartbeat acknowledgments to request-start monotonic time. Status and poll responses never renew the lease.
    public let leaseTtlSeconds: Int?
    public let leaseUntil: Double
    public let ownerEpoch: Int
    public let ownerId: Nullable<String>
    public let processes: [ProcessObservation]
    public let projectId: String
    public let protocolVersion: Nullable<String>
    public let providerId: String
    public let sessionId: String
    public let startupDeadline: Double
    public let status: AttachmentViewStatus
    public let uncertainOperations: [UncertainOperation]
    public let workspaceGeneration: String
    public let workspaceProvenance: WorkspaceProvenance
    /// Additional properties that are not explicitly defined in the schema
    public let additionalProperties: [String: JSONValue]

    public init(
        capabilities: [String],
        cleanupAt: Nullable<Double>,
        cleanupStatus: AttachmentViewCleanupStatus,
        environmentId: String,
        id: String,
        journalId: String,
        leaseTtlSeconds: Int? = nil,
        leaseUntil: Double,
        ownerEpoch: Int,
        ownerId: Nullable<String>,
        processes: [ProcessObservation],
        projectId: String,
        protocolVersion: Nullable<String>,
        providerId: String,
        sessionId: String,
        startupDeadline: Double,
        status: AttachmentViewStatus,
        uncertainOperations: [UncertainOperation],
        workspaceGeneration: String,
        workspaceProvenance: WorkspaceProvenance,
        additionalProperties: [String: JSONValue] = .init()
    ) {
        self.capabilities = capabilities
        self.cleanupAt = cleanupAt
        self.cleanupStatus = cleanupStatus
        self.environmentId = environmentId
        self.id = id
        self.journalId = journalId
        self.leaseTtlSeconds = leaseTtlSeconds
        self.leaseUntil = leaseUntil
        self.ownerEpoch = ownerEpoch
        self.ownerId = ownerId
        self.processes = processes
        self.projectId = projectId
        self.protocolVersion = protocolVersion
        self.providerId = providerId
        self.sessionId = sessionId
        self.startupDeadline = startupDeadline
        self.status = status
        self.uncertainOperations = uncertainOperations
        self.workspaceGeneration = workspaceGeneration
        self.workspaceProvenance = workspaceProvenance
        self.additionalProperties = additionalProperties
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.capabilities = try container.decode([String].self, forKey: .capabilities)
        self.cleanupAt = try container.decode(Nullable<Double>.self, forKey: .cleanupAt)
        self.cleanupStatus = try container.decode(AttachmentViewCleanupStatus.self, forKey: .cleanupStatus)
        self.environmentId = try container.decode(String.self, forKey: .environmentId)
        self.id = try container.decode(String.self, forKey: .id)
        self.journalId = try container.decode(String.self, forKey: .journalId)
        self.leaseTtlSeconds = try container.decodeIfPresent(Int.self, forKey: .leaseTtlSeconds)
        self.leaseUntil = try container.decode(Double.self, forKey: .leaseUntil)
        self.ownerEpoch = try container.decode(Int.self, forKey: .ownerEpoch)
        self.ownerId = try container.decode(Nullable<String>.self, forKey: .ownerId)
        self.processes = try container.decode([ProcessObservation].self, forKey: .processes)
        self.projectId = try container.decode(String.self, forKey: .projectId)
        self.protocolVersion = try container.decode(Nullable<String>.self, forKey: .protocolVersion)
        self.providerId = try container.decode(String.self, forKey: .providerId)
        self.sessionId = try container.decode(String.self, forKey: .sessionId)
        self.startupDeadline = try container.decode(Double.self, forKey: .startupDeadline)
        self.status = try container.decode(AttachmentViewStatus.self, forKey: .status)
        self.uncertainOperations = try container.decode([UncertainOperation].self, forKey: .uncertainOperations)
        self.workspaceGeneration = try container.decode(String.self, forKey: .workspaceGeneration)
        self.workspaceProvenance = try container.decode(WorkspaceProvenance.self, forKey: .workspaceProvenance)
        self.additionalProperties = try decoder.decodeAdditionalProperties(using: CodingKeys.self)
    }

    public func encode(to encoder: Encoder) throws -> Void {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try encoder.encodeAdditionalProperties(self.additionalProperties)
        try container.encode(self.capabilities, forKey: .capabilities)
        try container.encode(self.cleanupAt, forKey: .cleanupAt)
        try container.encode(self.cleanupStatus, forKey: .cleanupStatus)
        try container.encode(self.environmentId, forKey: .environmentId)
        try container.encode(self.id, forKey: .id)
        try container.encode(self.journalId, forKey: .journalId)
        try container.encodeIfPresent(self.leaseTtlSeconds, forKey: .leaseTtlSeconds)
        try container.encode(self.leaseUntil, forKey: .leaseUntil)
        try container.encode(self.ownerEpoch, forKey: .ownerEpoch)
        try container.encode(self.ownerId, forKey: .ownerId)
        try container.encode(self.processes, forKey: .processes)
        try container.encode(self.projectId, forKey: .projectId)
        try container.encode(self.protocolVersion, forKey: .protocolVersion)
        try container.encode(self.providerId, forKey: .providerId)
        try container.encode(self.sessionId, forKey: .sessionId)
        try container.encode(self.startupDeadline, forKey: .startupDeadline)
        try container.encode(self.status, forKey: .status)
        try container.encode(self.uncertainOperations, forKey: .uncertainOperations)
        try container.encode(self.workspaceGeneration, forKey: .workspaceGeneration)
        try container.encode(self.workspaceProvenance, forKey: .workspaceProvenance)
    }

    /// Keys for encoding/decoding struct properties.
    enum CodingKeys: String, CodingKey, CaseIterable {
        case capabilities
        case cleanupAt = "cleanup_at"
        case cleanupStatus = "cleanup_status"
        case environmentId = "environment_id"
        case id
        case journalId = "journal_id"
        case leaseTtlSeconds = "lease_ttl_seconds"
        case leaseUntil = "lease_until"
        case ownerEpoch = "owner_epoch"
        case ownerId = "owner_id"
        case processes
        case projectId = "project_id"
        case protocolVersion = "protocol_version"
        case providerId = "provider_id"
        case sessionId = "session_id"
        case startupDeadline = "startup_deadline"
        case status
        case uncertainOperations = "uncertain_operations"
        case workspaceGeneration = "workspace_generation"
        case workspaceProvenance = "workspace_provenance"
    }
}

File: rust/src/api/types/attachment_view.rs
pub use crate::prelude::*;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AttachmentView {
    #[serde(default)]
    pub capabilities: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    #[serde(with = "crate::core::number_serializers::option")]
    pub cleanup_at: Option<f64>,
    pub cleanup_status: AttachmentViewCleanupStatus,
    #[serde(default)]
    pub environment_id: String,
    #[serde(default)]
    pub id: String,
    #[serde(default)]
    pub journal_id: String,
    /// Renewal TTL; anchor only ready/heartbeat acknowledgments to request-start monotonic time. Status and poll responses never renew the lease.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub lease_ttl_seconds: Option<i64>,
    #[serde(default)]
    #[serde(with = "crate::core::number_serializers")]
    pub lease_until: f64,
    #[serde(default)]
    pub owner_epoch: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub owner_id: Option<String>,
    #[serde(default)]
    pub processes: Vec<ProcessObservation>,
    #[serde(default)]
    pub project_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub protocol_version: Option<String>,
    #[serde(default)]
    pub provider_id: String,
    #[serde(default)]
    pub session_id: String,
    #[serde(default)]
    #[serde(with = "crate::core::number_serializers")]
    pub startup_deadline: f64,
    pub status: AttachmentViewStatus,
    #[serde(default)]
    pub uncertain_operations: Vec<UncertainOperation>,
    #[serde(default)]
    pub workspace_generation: String,
    pub workspace_provenance: WorkspaceProvenance,
}

impl AttachmentView {
    pub fn builder() -> AttachmentViewBuilder {
        <AttachmentViewBuilder as Default>::default()
    }
}

#[derive(Clone, PartialEq, Default, Debug)]
#[non_exhaustive]
pub struct AttachmentViewBuilder {
    capabilities: Option<Vec<String>>,
    cleanup_at: Option<f64>,
    cleanup_status: Option<AttachmentViewCleanupStatus>,
    environment_id: Option<String>,
    id: Option<String>,
    journal_id: Option<String>,
    lease_ttl_seconds: Option<i64>,
    lease_until: Option<f64>,
    owner_epoch: Option<i64>,
    owner_id: Option<String>,
    processes: Option<Vec<ProcessObservation>>,
    project_id: Option<String>,
    protocol_version: Option<String>,
    provider_id: Option<String>,
    session_id: Option<String>,
    startup_deadline: Option<f64>,
    status: Option<AttachmentViewStatus>,
    uncertain_operations: Option<Vec<UncertainOperation>>,
    workspace_generation: Option<String>,
    workspace_provenance: Option<WorkspaceProvenance>,
}

impl AttachmentViewBuilder {
    pub fn capabilities(mut self, value: Vec<String>) -> Self {
        self.capabilities = Some(value);
        self
    }

    pub fn cleanup_at(mut self, value: f64) -> Self {
        self.cleanup_at = Some(value);
        self
    }

    pub fn cleanup_status(mut self, value: AttachmentViewCleanupStatus) -> Self {
        self.cleanup_status = Some(value);
        self
    }

    pub fn environment_id(mut self, value: impl Into<String>) -> Self {
        self.environment_id = Some(value.into());
        self
    }

    pub fn id(mut self, value: impl Into<String>) -> Self {
        self.id = Some(value.into());
        self
    }

    pub fn journal_id(mut self, value: impl Into<String>) -> Self {
        self.journal_id = Some(value.into());
        self
    }

    pub fn lease_ttl_seconds(mut self, value: i64) -> Self {
        self.lease_ttl_seconds = Some(value);
        self
    }

    pub fn lease_until(mut self, value: f64) -> Self {
        self.lease_until = Some(value);
        self
    }

    pub fn owner_epoch(mut self, value: i64) -> Self {
        self.owner_epoch = Some(value);
        self
    }

    pub fn owner_id(mut self, value: impl Into<String>) -> Self {
        self.owner_id = Some(value.into());
        self
    }

    pub fn processes(mut self, value: Vec<ProcessObservation>) -> Self {
        self.processes = Some(value);
        self
    }

    pub fn project_id(mut self, value: impl Into<String>) -> Self {
        self.project_id = Some(value.into());
        self
    }

    pub fn protocol_version(mut self, value: impl Into<String>) -> Self {
        self.protocol_version = Some(value.into());
        self
    }

    pub fn provider_id(mut self, value: impl Into<String>) -> Self {
        self.provider_id = Some(value.into());
        self
    }

    pub fn session_id(mut self, value: impl Into<String>) -> Self {
        self.session_id = Some(value.into());
        self
    }

    pub fn startup_deadline(mut self, value: f64) -> Self {
        self.startup_deadline = Some(value);
        self
    }

    pub fn status(mut self, value: AttachmentViewStatus) -> Self {
        self.status = Some(value);
        self
    }

    pub fn uncertain_operations(mut self, value: Vec<UncertainOperation>) -> Self {
        self.uncertain_operations = Some(value);
        self
    }

    pub fn workspace_generation(mut self, value: impl Into<String>) -> Self {
        self.workspace_generation = Some(value.into());
        self
    }

    pub fn workspace_provenance(mut self, value: WorkspaceProvenance) -> Self {
        self.workspace_provenance = Some(value);
        self
    }

    /// Consumes the builder and constructs a [`AttachmentView`].
    /// This method will fail if any of the following fields are not set:
    /// - [`capabilities`](AttachmentViewBuilder::capabilities)
    /// - [`cleanup_status`](AttachmentViewBuilder::cleanup_status)
    /// - [`environment_id`](AttachmentViewBuilder::environment_id)
    /// - [`id`](AttachmentViewBuilder::id)
    /// - [`journal_id`](AttachmentViewBuilder::journal_id)
    /// - [`lease_until`](AttachmentViewBuilder::lease_until)
    /// - [`owner_epoch`](AttachmentViewBuilder::owner_epoch)
    /// - [`processes`](AttachmentViewBuilder::processes)
    /// - [`project_id`](AttachmentViewBuilder::project_id)
    /// - [`provider_id`](AttachmentViewBuilder::provider_id)
    /// - [`session_id`](AttachmentViewBuilder::session_id)
    /// - [`startup_deadline`](AttachmentViewBuilder::startup_deadline)
    /// - [`status`](AttachmentViewBuilder::status)
    /// - [`uncertain_operations`](AttachmentViewBuilder::uncertain_operations)
    /// - [`workspace_generation`](AttachmentViewBuilder::workspace_generation)
    /// - [`workspace_provenance`](AttachmentViewBuilder::workspace_provenance)
    pub fn build(self) -> Result<AttachmentView, BuildError> {
        Ok(AttachmentView {
            capabilities: self
                .capabilities
                .ok_or_else(|| BuildError::missing_field("capabilities"))?,
            cleanup_at: self.cleanup_at,
            cleanup_status: self
                .cleanup_status
                .ok_or_else(|| BuildError::missing_field("cleanup_status"))?,
            environment_id: self
                .environment_id
                .ok_or_else(|| BuildError::missing_field("environment_id"))?,
            id: self.id.ok_or_else(|| BuildError::missing_field("id"))?,
            journal_id: self
                .journal_id
                .ok_or_else(|| BuildError::missing_field("journal_id"))?,
            lease_ttl_seconds: self.lease_ttl_seconds,
            lease_until: self
                .lease_until
                .ok_or_else(|| BuildError::missing_field("lease_until"))?,
            owner_epoch: self
                .owner_epoch
                .ok_or_else(|| BuildError::missing_field("owner_epoch"))?,
            owner_id: self.owner_id,
            processes: self
                .processes
                .ok_or_else(|| BuildError::missing_field("processes"))?,
            project_id: self
                .project_id
                .ok_or_else(|| BuildError::missing_field("project_id"))?,
            protocol_version: self.protocol_version,
            provider_id: self
                .provider_id
                .ok_or_else(|| BuildError::missing_field("provider_id"))?,
            session_id: self
                .session_id
                .ok_or_else(|| BuildError::missing_field("session_id"))?,
            startup_deadline: self
                .startup_deadline
                .ok_or_else(|| BuildError::missing_field("startup_deadline"))?,
            status: self
                .status
                .ok_or_else(|| BuildError::missing_field("status"))?,
            uncertain_operations: self
                .uncertain_operations
                .ok_or_else(|| BuildError::missing_field("uncertain_operations"))?,
            workspace_generation: self
                .workspace_generation
                .ok_or_else(|| BuildError::missing_field("workspace_generation"))?,
            workspace_provenance: self
                .workspace_provenance
                .ok_or_else(|| BuildError::missing_field("workspace_provenance"))?,
        })
    }
}
